libunwind/src/DwarfInstructions.hpp

```diff
@@ -365,7 +365,12 @@ int DwarfInstructions<A, R>::stepWithDwarf(A &addressSpace, pint_t pc,
 
       // Return address is address after call site instruction, so setting IP to
       // that does simulates a return.
-      newRegisters.setIP(returnAddress);
+      //
+      // In case of this is frame of signal handler, the IP should be
+      // incremented, because the IP saved in the signal handler points to
+      // first non-executed instruction, while FDE/CIE expects IP to be after
+      // the first non-executed instruction.
+      newRegisters.setIP(returnAddress + cieInfo.isSignalFrame);
```

Comment:
The idea here is that `cieInfo.isSignalFrame` is a boolean and will increment by 1. However, how do we ensure that the IP should be incremented by 1 and not say 4 or some other size for the instruction? Particularly in the case of an ARMv7 + thumb2 environment, this can potentially change the meaning as the bottom bit is used as an ISA selector.

Reply:
>The idea here is that cieInfo.isSignalFrame is a boolean and will increment by 1

Correct

>However, how do we ensure that the IP should be incremented by 1 and not say 4 or some other size for the instruction?

This is only needed to execute proper CFI here - https://github.com/llvm/llvm-project/blob/23d1d959583c35e6eab7e3e70d4c16449b418563/libunwind/src/DwarfParser.hpp#L433-L437

So 1 is enough

Reply:
Could the register state not being used for anything else in between and thus be misleading?

Reply:
I guess I'm thinking about someone using libunwind and manually walking the FDE. The single step alters the register state (`.setIP`). As long as you are just walking the FDE for unwinding as we do it is fine, but were someone to do something non-standard in between, it would just give you an unaligned IP.

Reply:
OK, I see, thanks.
I can move this logic into `DwarfInstructions<A, R>::stepWithDwarf`, but still, the IP will be different in case of regular unwind (after first non executed instruction) and unwind from signal (before first non-executed instruction)

Reply:
Right, I think that part is reasonable. When you unwind from a signal, the IP should point to the next instruction vs unwinding from a regular context.

Reply:
So you think moving this logic into `stepWithDwarf` will not cause any concerns?

Reply:
@compnerd I've moved this logic into `UnwindCursor<A, R>::setInfoBasedOnIPRegister`, and verified it with the initial reproducer (it is too complex to convert it to a test) - the patch works correctly